include/triton/backend/device_memory_tracker.h

```diff
@@ -0,0 +1,150 @@
+// Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <map>
+#include <memory>
+#include <mutex>
+#include <unordered_map>
+#include <cupti.h>
+
+static_assert(
+    sizeof(uint64_t) >= sizeof(uintptr_t),
+    "The implementation is storing address pointer as uint64_t, "
+    "must ensure the space for pointer is <= sizeof(uint64_t).");
+
+namespace triton { namespace backend {
+
+/// DeviceMemoryTracker is a backend utility provided to track the memory
+/// allocated for a particular model and associated model instance. This utility
+/// is often used for backend that set
+/// TRITONBACKEND_BackendAttributeSetEnableMemoryTracker to true, which provides
+/// additional information to Triton for making decision on model scaling and
+/// deployment.
+///
+/// Caveat: The memory tracker is implemented with CUPTI library which currently
+/// only supports single client/subscriber. This is an known limitation and as a
+/// result, the memory tracker can cause unexpected application failure if other
+/// component of the Triton process also uses CUPTI with a different
+/// configuration, for example, the framework used by the backend may have
+/// implemented similar profiler with CUPTI. Therefore, before enabling this
```

Comment:
Is it possible for the DeviceMemoryTracker internally to check if there are other CUPTI clients active via some CUPTI APIs?

Reply:
See https://github.com/triton-inference-server/backend/pull/80/files#diff-f8ad565984e1abf1676a6d0b85f921cb0a29f78f6532debe29d2ca8d110dd51dR108-R127